pkg/controllers/dynakube/extension/statefulset.go

```diff
@@ -0,0 +1,387 @@
+	if !r.dk.PrometheusEnabled() {
+		if meta.FindStatusCondition(*r.dk.Conditions(), extensionsControllerStatefulSetConditionType) == nil {
+			return nil
+		}
+		defer meta.RemoveStatusCondition(r.dk.Conditions(), extensionsControllerStatefulSetConditionType)
+
+		sts, err := statefulset.Build(r.dk, statefulsetName)
+		if err != nil {
+			log.Error(err, "could not build "+statefulsetName+" during cleanup")
+
+			return err
+		}
+
+		err = statefulset.Query(r.client, r.apiReader, log).Delete(ctx, sts)
+
+		if err != nil {
+			log.Error(err, "failed to clean up "+statefulsetName+" statufulset")
+
+			return nil
+		}
+
+		return nil
+	}
+
+	if r.dk.Status.ActiveGate.ConnectionInfoStatus.TenantUUID == "" {
+		conditions.SetStatefulSetOutdated(r.dk.Conditions(), extensionsControllerStatefulSetConditionType, statefulsetName)
+		return errors.New("tenantUUID unknown")
+	}
+
+	return r.createOrUpdateStatefulset(ctx)
+}
+
+func (r *reconciler) createOrUpdateStatefulset(ctx context.Context) error {
+	desiredSts, err := statefulset.Build(r.dk, statefulsetName,
+		setReplicas(),
+		setPodManagementPolicy(),
+		setLabels(r.dk.Spec.Templates.ExtensionExecutionController.Labels, r.dk.Name),
+		setAnnotations(r.dk.Spec.Templates.ExtensionExecutionController.Annotations),
+		setAffinity(),
+		setTolerations(r.dk.Spec.Templates.ExtensionExecutionController.Tolerations),
+		setTopologySpreadConstraints(r.dk.Spec.Templates.ExtensionExecutionController.TopologySpreadConstraints, r.dk.Name),
+		setTlsRef(r.dk.Spec.Templates.ExtensionExecutionController.TlsRefName),
+		setImagePullSecrets(r.dk.ImagePullSecretReferences()),
+		setVolumes(r.dk.Name, r.dk.Spec.Templates.ExtensionExecutionController.PersistentVolumeClaim),
+		setContainer(),
+		setContainerImage(r.dk.Spec.Templates.ExtensionExecutionController.ImageRef.Repository, r.dk.Spec.Templates.ExtensionExecutionController.ImageRef.Tag),
+		setContainerResources(r.dk.Spec.Templates.ExtensionExecutionController.Resources),
+		setContainerEnvs(r.dk.Name, r.dk.Namespace, r.dk.Status.ActiveGate.ConnectionInfoStatus.TenantUUID, "activegate", r.dk.Status.KubeSystemUUID),
+		setContainerVolumeMounts(),
+	)
+
+	if err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), extensionsControllerStatefulSetConditionType, err)
+		return err
+	}
+
+	if err := setHash(desiredSts); err != nil {
+		conditions.SetKubeApiError(r.dk.Conditions(), extensionsControllerStatefulSetConditionType, err)
+		return err
+	}
+
+	_, err = statefulset.Query(r.client, r.apiReader, log).WithOwner(r.dk).CreateOrUpdate(ctx, desiredSts)
+	if err != nil {
+		log.Info("failed to create/update " + statefulsetName + " statefulset")
+		conditions.SetKubeApiError(r.dk.Conditions(), extensionsControllerStatefulSetConditionType, err)
+
+		return err
+	}
+
+	conditions.SetStatefulSetCreated(r.dk.Conditions(), extensionsControllerStatefulSetConditionType, desiredSts.Name)
+
+	return nil
+}
+
+func setReplicas() func(o *appsv1.StatefulSet) {
```

Comment:
must be part of the framework, so we can reuse it. So please move it to builder.go, check how I did service it will help you to inspire.